### PR TITLE
Remove project name from schema

### LINKED ISSUE #541: Removing signac project id
I've been meaning to raise this question for a while, and in anticipation of resolving #527 I think now is a good time to consider it: should we remove the project name in signac schema version 2? Currently we use it sparingly, pretty much just to provide a string representation of a project and in associated functions like naming a temporary project. signac-flow uses this to generate the unique submission id, but the fact that a project name is not guaranteed to be unique anyway forced us to add the project directory to the submission id hash anyway. We can always introduce alternative methods in signac-flow's FlowProject to give more meaningfully named submissions, if the main value is that a submitted cluster job can be associated with a project based on the name. Is there any other meaningful upside to keeping this, rather than just falling back to the name of the current directory? I think it's worth considering as part of the shift, especially since we're removing workspace configurability.

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This PR updates the v1 to v2 schema migration to remove project names and store preexisting ones in the project document.

## Motivation and Context

Resolves #541. Part of the changeset of #643.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [ ] New feature
- [x] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added all related issue and pull request numbers for future reference (if applicable). See example below.

### CODE REVIEW

**Comment by vyasr** [comment]
@bdice @csadorf the more I work with these schema-related PRs the more I wonder about the value in using semantic versioning-style versions for schemas. I don't think they're quite appropriate, and I think they introduce more complexity than is worthwhile, especially since there are no equivalents to "patches" AFAICT. I could theoretically imagine a schema version bump that would not _require_ migrating from an old schema by adding support for optional keys, and that could constitute a minor version, but is that important for us to really support? I think the schema code would be significantly simplified by just using a simple integer-indexed version scheme, and that is practically what we are doing right now anyway. Thoughts?

**Comment by bdice** [comment]
@vyasr My recollection is that we will only make “major” (breaking) schema versions, but we use PEP 440 version identifiers (not necessarily semantic versioning) just because it gives us flexibility should we need it.

**Comment by vyasr** [comment]
I mainly ask because the migration code ends up doing a lot of conversion back and forth between "proper" versions and integers (or really, a string containing a single number) because all of the stored variables use version numbers like "1". So I think our options are to either lean into the version numbers by replacing all of those with `packaging.Version` objects or to move completely away from them in favor of simple integers. @csadorf do you have a preference? I wasn't part of the early design discussions around the migration logic so I'm not sure if you two had specific thoughts at the time on the importance of full versions.

**Comment by vyasr** [comment]
https://github.com/glotzerlab/signac/pull/684/commits/e1e9944f2407253335de206894b2e854cbaed2aa addresses https://github.com/glotzerlab/signac/pull/643#discussion_r776492146. That case isn't quite covered by the tests added in #682 since that also requires the global config file to be missing, but I think this is good enough for now. I had to change a test because in fact the behavior of various error cases handled by the config CLI wasn't consistent. The existing config code actually had some errors in it that were masked by old behaviors and were uncovered by this change, which is why we observed this problem in the first place. We should aim to improve testing of the config CLI further, but I think that's out of scope for this PR as long as we've resolved the current issues.

**Comment by csadorf** [comment]
@vyasr Yes, I'm in favor of just going with simple integer increments. While schemas can generally be backwards-compatible, I can't imagine a scenario where communicating compatibility through the version number is adding much value in this specific context. We can always "manually" document specific compatiblity ranges if there is a need for this.

File: signac/__main__.py
Comment (vyasr) [reply signac/__main__.py:725]: Both `read_config` and `load_config` are guaranteed to return a `Config` object if they succeed. This change is necessary because `load_config` can return an empty Config if no config files were found. Previously, you would get a printed log (although not an exception) saying that the "project" key was missing as you pointed out [here](https://github.com/glotzerlab/signac/pull/643#discussion_r776492146). That can no longer happen since the key doesn't exist, and we can't replace it with "project_dir" since I plan to remove that from the config as well (in the next PR where I change load_config to return a pair of values). Therefore, the two possibilities are that `cfg is None` -- which means that the user specified `args.local` and the project config didn't exist -- or that `cfg == Config({})`, which can be evaluated as falsy.

File: signac/__main__.py
Comment (vyasr) [reply signac/__main__.py:759]: If you look up 10 lines to 749 you'll see that is an error case. This block was always unreachable.

Comment (vyasr) [reply signac/__main__.py:759]: Also, just so it's clear, this won't change the case where a user just specified local or globalcfg. If the user specified _neither_ however, the expected message is actually that neither file was found (because we search both by default), not an empty string (as was previously shown).

File: signac/contrib/migration/__init__.py
Comment (vyasr) [reply signac/contrib/migration/__init__.py:48]: Yup, I'll merge those changes down after I merge that PR, I'll need to update this PR anyway before tests pass (on oldest).

File: tests/test_project.py
Comment (vyasr) [reply tests/test_project.py:101]: Well it would help if this test actually contained an assertion... 😂 

Looks like this test has remained unchanged since https://github.com/glotzerlab/signac/commit/bbf7e7b0db9164bf10efaa84ab2888a8bd15e307 way back in 2016. Probably good to enable it now.

Comment (bdice) [reply tests/test_project.py]: > To know oneself, one should assert oneself.
> -- Albert Camus

File: signac/__main__.py
Comment (bdice) [inline signac/__main__.py:733]: Should this exit with some nonzero return code? I notice that some of our code paths calling `_print_err` are informational, while others indicate actual errors and are followed by `sys.exit(1)`. Some code paths raise `RuntimeWarning`, which also exits with code 1. https://github.com/glotzerlab/signac/blob/c8cc307566c78f9f960abc43fb793b6b368ec18d/signac/__main__.py#L1715-L1719

Perhaps we need to assess this in a separate PR, and make sure all the code in `__main__.py` is consistent about using nonzero exit codes for errors. For now I'll mark a couple cases that are obvious.

```suggestion
        sys.exit(1)
```

File: signac/__main__.py
Comment (bdice) [inline signac/__main__.py:766]:
```suggestion
        sys.exit(1)
```

File: signac/contrib/project.py
Comment (bdice) [inline signac/contrib/project.py:167]: For a `Job`, we use `job.id` as the result of `str(job)`. That is, the string form is a simple unique identifier. I think the analogous thing here is just the project's root directory? I'm happy with either solution, but I wanted to pose that idea.
```suggestion
        return str(self.root_directory())
```

**Review by bdice — APPROVED** [review]
A few more thoughts. I'm fine if you merge this once you've considered/applied my comments, so I'm approving.

File: signac/contrib/project.py
Comment (vyasr) [reply signac/contrib/project.py:167]: I agree, that's probably better. The repr is a bit uglier and more verbose than is really needed.

**Pull request merged**
